File: fin/api/routers/installments.py
"""Instalment plans and outstanding liability."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException

from ... import db as dbm
from ..deps import get_conn

router = APIRouter(tags=["installments"])


@router.get("/installments")
def list_plans(active_only: bool = False, conn=Depends(get_conn)) -> dict:
    """Plans with progress and what is still owed.

    Outstanding totals are grouped by currency. Summing a HKD plan and a USD
    plan into one liability figure would be meaningless.
    """
    plans = dbm.load_installment_plans(conn, active_only=active_only)

    outstanding: dict[str, int] = {}
    monthly: dict[str, int] = {}
    for p in plans:
        if p["status"] != "active":
            continue
        ccy = p["outstanding"]["currency"]
        outstanding[ccy] = outstanding.get(ccy, 0) + p["outstanding"]["amount"]
        monthly[ccy] = monthly.get(ccy, 0) + p["per_installment"]["amount"]

    return {
        "plans": plans,
        "outstanding_by_currency": [
            {"currency": c, "amount": a} for c, a in sorted(outstanding.items())],
        # Remaining instalments are known future outflows — the only genuinely
        # predictable part of a spending forecast.
        "committed_monthly_by_currency": [
            {"currency": c, "amount": a} for c, a in sorted(monthly.items())],
    }


@router.get("/installments/{plan_id}")
def get_plan(plan_id: str, conn=Depends(get_conn)) -> dict:
    plans = dbm.load_installment_plans(conn)
    plan = next((p for p in plans if p["id"] == plan_id), None)
    if plan is None:
        raise HTTPException(404, "no such plan")
    plan["charges"] = [dict(r) for r in conn.execute(
        "SELECT id, txn_date, description_raw, amount_booked, currency_booked, "
        "       installment_seq, "
        "       CASE WHEN installment_seq IS NULL THEN TRUE ELSE FALSE END AS is_settlement "
        "FROM txn WHERE installment_plan_id=%s AND duplicate_of_id IS NULL "
        "ORDER BY txn_date, installment_seq NULLS LAST", (plan_id,))]
    return plan

File: web/src/app/features/installments/installments.ts
import { Component, inject, signal } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { Api } from '../../core/api.service';
import { MoneyPipe, ShortDatePipe } from '../../core/money.pipe';
import { InstallmentPlan, Money } from '../../core/models';

interface ScheduleRow {
  key: string;
  seq: number | null;
  date: string;
  amount: Money;
  paid: boolean;
  settlement?: boolean;
  description?: string;
  txnId?: string;
}

/**
 * Instalment plans.
 *
 * The ledger stores what actually hit the account, one charge at a time — that
 * is forced, because the balance check proves we captured every transaction and
 * an accrual row would break it. The consequence is that the two numbers that
 * matter about a plan, what you still owe and what is already committed for next
 * month, exist nowhere in the rows themselves. This page is where they live.
 *
 * Outstanding is never summed across currencies. A HKD plan and a USD plan do
 * not add up to a liability figure.
 */
@Component({
  selector: 'app-installments',
  imports: [FormsModule, MoneyPipe, ShortDatePipe],
  templateUrl: './installments.html',
  styleUrl: './installments.css',
})
export class InstallmentsPage {
  private api = inject(Api);

  loading = signal(true);
  activeOnly = signal(true);
  plans = signal<InstallmentPlan[]>([]);
  outstanding = signal<Money[]>([]);
  monthly = signal<Money[]>([]);
  expanded = signal<string | null>(null);
  detail = signal<InstallmentPlan | null>(null);

  constructor() {
    this.load();
  }

  load(): void {
    this.loading.set(true);
    this.api.installments(this.activeOnly()).subscribe({
      next: (res) => {
        this.plans.set(res.plans);
        this.outstanding.set(res.outstanding_by_currency);
        this.monthly.set(res.committed_monthly_by_currency);
        this.loading.set(false);
      },
      error: () => this.loading.set(false),
    });
  }

  setActiveOnly(value: boolean): void {
    this.activeOnly.set(value);
    this.load();
  }

  toggle(plan: InstallmentPlan): void {
    if (this.expanded() === plan.id) {
      this.expanded.set(null);
      return;
    }
    this.expanded.set(plan.id);
    this.detail.set(null);
    this.api.installment(plan.id).subscribe({ next: (p) => this.detail.set(p) });
  }

  progress(plan: InstallmentPlan): string {
    if (plan.status === 'completed') return '100%';
    return `${(plan.paid_count / Math.max(1, plan.term_months)) * 100}%`;
  }

  /**
   * The full term: charges we hold, then the ones still to come.
   *
   * Remaining instalments are the only genuinely predictable part of a spending
   * forecast — the amount and the date are both already agreed.
   */
  schedule(plan: InstallmentPlan): ScheduleRow[] {
    const charges = plan.charges ?? [];
    const bySeq = new Map<number, (typeof charges)[number]>();
    charges.filter((c) => c.installment_seq != null)
      .forEach((c) => bySeq.set(c.installment_seq!, c));

    const rows: ScheduleRow[] = [];
    for (let seq = 1; seq <= plan.term_months; seq++) {
      const charge = bySeq.get(seq);
      if (charge) {
        rows.push({
          key: `charge-${charge.id}`,
          seq,
          date: charge.txn_date,
          amount: { amount: charge.amount_booked, currency: charge.currency_booked },
          paid: true,
          description: charge.description_raw,
          txnId: charge.id,
        });
      } else {
        if (plan.status === 'completed') continue;
        rows.push({
          key: `due-${seq}`,
          seq,
          date: addMonths(plan.start_date, seq - 1),
          amount: plan.per_installment,
          paid: false,
        });
      }
    }
    for (const charge of charges.filter((c) => c.is_settlement)) {
      rows.push({
        key: `settlement-${charge.id}`,
        seq: null,
        date: charge.txn_date,
        amount: { amount: charge.amount_booked, currency: charge.currency_booked },
        paid: true,
        settlement: true,
        description: charge.description_raw,
        txnId: charge.id,
      });
    }
    return rows;
  }

  /** Below this, the grouping was a judgement call rather than a certainty. */
  isUncertain(plan: InstallmentPlan): boolean {
    return !plan.is_confirmed && plan.confidence < 0.95;
  }
}

/** Same day-of-month, clamped to the length of the target month. */
function addMonths(iso: string, months: number): string {
  const [y, m, d] = iso.slice(0, 10).split('-').map(Number);
  const target = new Date(y, m - 1 + months, 1);
  const lastDay = new Date(target.getFullYear(), target.getMonth() + 1, 0).getDate();
  target.setDate(Math.min(d, lastDay));
  const pad = (n: number) => String(n).padStart(2, '0');
  return `${target.getFullYear()}-${pad(target.getMonth() + 1)}-${pad(target.getDate())}`;
}
